### DISCUSSION TITLE
Forward on parse Error in Json::from_data instead of failing

### DISCUSSION THREAD

**Discussion by LorenzSchueler** [rwf2/Rocket#1830] [Questions] [answered]
Hello,
I'm using rocket v0.5 and tried to use multiple routes to the same URL. When using different parameters this works perfectly. However I also tried to use this with POST request where the payload can be a Json of type a or type b. The problem is that the FromData implementation for Json returns an Outcome::Failure when it encounters a Error::Parse instead of forwarding it. 
I thought I could just write my own FromData implementation which internally uses the FromData implementation of Json but converts an Outcome::Failure with Status::UnprocessableEntity to an  Outcome::Forward. the problem here is that from_data consumes data but data is not Clone so that I could put it into Outcome::Forward.
I'm wondering if I'm missing something or if this is intended because cloning data could be expensive.

**Comment by jebrosen** [comment, marked as answer]
> When using different parameters this works perfectly. However I also tried to use this with POST request where the payload can be a Json of type a or type b.

Since Rocket's JSON support is based on `serde_json`, one way to accomplish a single route with multiple kinds of payload would be to use serde's support for several different [enum representations](https://serde.rs/enum-representations.html) to have `serde` determine which type to deserialize based on the data.

> I'm wondering if I'm missing something or if this is intended because cloning data could be expensive.

Yes, I think this is more or less the same as issue #775. I explained the reason for this limitation in https://github.com/SergioBenitez/Rocket/issues/775#issuecomment-631590955, along with a few potential solutions for the future. But, none of them will be ready for 0.5 - and they would likely require parsing the same data multiple times, making it a slower option than `serde` anyway.
